### ISSUE TITLE
MDRenewMode always: cert renewal on each server reload

### ISSUE THREAD

**Issue by middlebrain** [icing/mod_md#163]
With ```MDRenewMode always``` Managed Domain certificates will be renewed on each ```apachectl graceful```.  In my opinion, ```renewal``` should only be done when the expiry rules apply.

**Comment by icing** [comment]
That is certainly how `MDRenewMode always` is intended to work. Do you have a log of such a run, ideally at `LogLevel md:trace1` and the minimal config?

**Comment by middlebrain** [comment]
Here it is:

The following instructions are for a test domain ```mydomain.net``` and ```boulder``` as CA. Virtual hosts are not required for this demonstration.

Base config:
```
MDCertificateAgreement accepted
MDCertificateAuthority http://boulder:4001/directory
MDRenewMode always
```
To get a "third-party" certificate:
```
MDomain mydomain.net
```
Reload Apache to get a the cert from ```boulder```:
```
apachectl graceful
```
Reload Apache a second time to install the cert:
```
apachectl graceful
```
Copy ```MDStoreDir/mydomain.net/{pubcert.pem,privkey.pem}``` to ```/etc/apache2/ssl.*/...``` (for example) and change the md config to
```
# MDomain mydomain.net

<MDomain test.mydomain.net>
  MDCertificateFile /etc/apache2/ssl.crt/mydomain.net/pubcert.pem
  MDCertificateKeyFile /etc/apache2/ssl.key/mydomain.net/privkey.pem
</MDomain>
```
What I found was, that the hostname in the ```<MDomain ...>``` directive must be somehow different from the hostname of the ```MDomain``` directive for the "third-party" certificate. When the hostnames are the same, ```renewal``` is skipped (in ```md_reg.c``` line ```225``` is executed).

Anyway, next step: Remove all in the ```MDStore``` directory except ```accounts``` and ```md_store.json```, then reload Apache to get the cert for ```test.mydomain.net```:
```
apachectl graceful
```
Up to this point, everything is still fine. And now reload Apache again to install the cert
```
apachectl graceful
```
and see my comments in the following ```md:trace1``` log:
```
# install cert and remove staging directory:

md_acme_drive.c(503): test.mydomain.net: init_base driver
md_store_fs.c(668): start remove of md staging/test.mydomain.net/order.json
md_store_fs.c(974): using archive dir: /var/lib/apache2/md/archive/test.mydomain.net.1
md_reg.c(1135): test.mydomain.net: load done
mod_md.c(693): cecking duplicate ssl assignments
mod_md.c(481): md[test.mydomain.net]: auto add domains
md_acme_drive.c(503): test.mydomain.net: init_base driver
md_acme_drive.c(580): test.mydomain.net: init driver
mod_md_drive.c(298): md(test.mydomain.net): state=1, created drive job

# At this point the install job is done and the following
# should not happen in my opinion.

# A new staging job is created and executed.

md_acme_drive.c(503): test.mydomain.net: init_base driver
md_acme_drive.c(580): test.mydomain.net: init driver
md_acme_drive.c(624): test.mydomain.net: reset staging area, will
md_curl.c(394): request <-- 200
md_acme.c(676): directory lookup response: 200
md_acme.c(460): add acme POST: http://boulder:4001/acme/acct/51
md_curl.c(394): request <-- 200
md_acme.c(257): acme payload(len=2): {}
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme.c(460): add acme POST: http://boulder:4001/acme/new-order
md_acme.c(257): acme payload(len=60): {"identifiers":[{"type":"dns","value":"test.mydomain.net"}]}
md_curl.c(394): request <-- 201
md_acme.c(294): response: 201
md_acme_order.c(308): new order at http://boulder:4001/acme/order/51/394
md_acme.c(483): add acme GET: http://boulder:4001/acme/order/51/394
md_acme.c(257): acme payload(len=0): 
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme.c(483): add acme GET: http://boulder:4001/acme/authz-v3/634
md_acme.c(257): acme payload(len=0): 
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme.c(483): add acme GET: http://boulder:4001/acme/authz-v3/634
md_acme.c(257): acme payload(len=0): 
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme.c(483): add acme GET: http://boulder:4001/acme/order/51/394
md_acme.c(257): acme payload(len=0): 
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme.c(460): add acme POST: http://boulder:4001/acme/finalize/51/394
md_acme.c(257): acme payload(len=890): {"csr":"MIICkDCCAXgCAQAwHDEaMBgGA1UEAwwRdGVzdC5teWRvbWFpbi5uZXQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDUW_So5UF0j6NWYn2MOReUDRLnrpfeod_gJqlUZp7AJR5NNnzCJZrTFQVFXefXBNl_bHi6xGTyHLHhBi12zm02fs1oAJN_jWyR5-bXnbBCQdYmBJd18-Wy_bO3Zi348Ejm38mGv_IcrY75K_sBitXQtiRqQmKIuTOUgeYxIHkB1ROF7PFpPlkYJecDQaUcnbEr2r3qTPL6xriO97S4SgpGS6b_UbudvZV_S_7tnKzC8kS2FX2ZzJxSDOkTeUjNkItKXSfAHZFbVMvWOpkAP4orH3Wz-hAC67iUF67pxkiZywGaPZFZoNOZ3NqULp3Q9or8jVAa5gODHKsYIodSC3z9AgMBAAGgLzAtBgkqhkiG9w0BCQ4xIDAeMBwGA1UdEQQVMBOCEXRlc3QubXlkb21haW4ubmV0MA0GCSqGSIb3DQEBCwUAA4IBAQC3OLWVmzkeFFt1kSepVIbh9JlYhUfMmUu8UUq2eM0jC4RKb6HYIBzGKlNXQ0pr8BBtel-y81OlQmq8b1FLX_lYrG159LmcofDXLz8vlREg9-yRrt7ASek46A9IqcnZ_PkAuRFP74aiRG_XPLuFxmm1q7ljOQGcwl97J0Upa45Co4HzMWAjoZXbPv2WGLNnibXqu1aNxpoJrJwNX9JVvmdh7Xz8-P9-srz1Y2g8EGT5KpIQXpd8Z2ju3rXvqu68bu4CsZpdbFJSOE_CTjSh63OGWvnOH68yoXZEUZ1IrJR_mdCLIo2kfjeMN_BtDLZCCwtS30RGLF49JyCmuhm4lfrC"}
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme.c(483): add acme GET: http://boulder:4001/acme/order/51/394
md_acme.c(257): acme payload(len=0): 
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme_drive.c(243): retrieving cert from http://boulder:4001/acme/cert/ff7979182acfc75612443a243e23a3fa06a5
md_acme.c(483): add acme GET: http://boulder:4001/acme/cert/ff7979182acfc75612443a243e23a3fa06a5
md_acme.c(257): acme payload(len=0): 
md_curl.c(394): request <-- 200
md_acme.c(294): response: 200
md_acme_drive.c(438): got chain with 2 certs (0. attempt)
md_store_fs.c(668): start remove of md staging/test.mydomain.net/order.json
md_acme_drive.c(748): test.mydomain.net: certificate ready, activation delay set to 1d
mod_md_drive.c(156): test.mydomain.net: saving job props
```
With each subsequent ```apachectl graceful``` a new staging job is executed.

However, I may misunderstood the meaning of the ```MDRenewMode always``` directive.

**Comment by icing** [comment]
I am not sure I understand what you want to achieve with this. Maybe you need to explain more on where you start and want to end up.

The `MDRenewMode` controls if the module should watch and renew certificates for a `MDomain`.  The default is `auto` and means: "check if the MDomain is used in a `VirtualHost` somewhere and only then care about renewing certs. If it is not in use, do not care.

But some people want certificates from Lets Encrypt to use somewhere else. For example, you might want to have a certificate for your IMAP server, but `imap.mydomain.net` is not a site in your apache server. Then you would configure:

```
<MDomain imap.mydomain.net>
  MDRenewMode always
</MDomain>
```
and the module would get/renew a cert for that.

The third value is `manual`, meaning just install the certificate in the appropriate `VirtualHost`s, but do not care about renewing them. This mode is implicitly enabled when you configure static files via `MDCertificateFile`.

Since you seem to do a mix of all this, I might not have explained it well enough in the documentation....

**Comment by middlebrain** [comment]
My core problem is that the instructions at https://github.com/icing/mod_md#how-to-migrate-a-https-host do not work as hoped:

As soon as the new certificate is installed (after the second ```apachectl graceful```), it is used immediately. This behaviour is too delicate for me, especially since I want to convert a whole bunch of virtual hosts.

In my opinion, the new certificate should only be used after ```SSLCertificateFile``` and ```SSLCertificateKeyFile``` have been removed from the virtual host configuration.

Looking for an alternative solution I found https://github.com/icing/mod_md#how-to-use-other-certificates together with the ```MDRenewMode always``` statement.
